File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character,Character> map = new HashMap<>(); 
        map.put('}','{');
        map.put(']','[');
        map.put(')','(');

        //Push the string onto the stack
        for(char c : s.toCharArray()){
            //Check if you get a closed bracket
            if(map.containsKey(c)){
                //if the stack is empty or the top elem in stack != the open bracket it is invalid
                if(stack.isEmpty() || stack.pop() != map.get(c)){
                    return false;
                }
            }else{
                stack.push(c);
            }   
        }
        // if the stack is empty, then that means it is valid parentheses
        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        """
        we can use a stack when we iterate through the string to process the closing bracket
        hashmap to map open bracket to close. this helps us process the most recent bracket

        the open bracket either must have another open bracket within it or the same closing bracket

        ([])
        ^

        stack = [ (, [ ]
        """

        hashmap = {'(' : ')', '{' : '}', '[' : ']'}

        stack = []
        for char in s:
            # check if current character is an open bracket, add to stack and continue to next char
            if char in "([{":
                stack.append(char)
                continue

            if not stack:
                return False
            
            if char not in hashmap[stack.pop()]: # check if next char is closed bracket for current open bracket
                return False
            
        # return true if we were able to process all parentheses
        return len(stack) == 0


            
